src/theory/arith/nl/piand_solver.cpp

```diff
@@ -170,7 +170,182 @@ void PIAndSolver::checkInitialRefine()
   }
 }
 
-void PIAndSolver::checkFullRefine() {}
+void PIAndSolver::checkFullRefine()
+{
+  NodeManager* nm = nodeManager();
+  Trace("piand-check") << "PIAndSolver::checkFullRefine";
+  Trace("piand-check") << "PIAND terms: " << std::endl;
+  for (const std::pair<const Node, std::vector<Node> >& is : d_piands)
+  {
+    int index = 0;
+    for (const Node& i : is.second)
+    {
+      index++;
+      Node valAndXY = d_model.computeAbstractModelValue(i);
+      Node valAndXYC = d_model.computeConcreteModelValue(i);
+
+      Node k = i[0];
+      Node x = i[1];
+      Node y = i[2];
+      Node valK = d_model.computeConcreteModelValue(k);
+      Node valX = d_model.computeConcreteModelValue(x);
+      Node valY = d_model.computeConcreteModelValue(y);
+
+      Integer model_piand = valAndXYC.getConst<Rational>().getNumerator();
+      Integer model_k = valK.getConst<Rational>().getNumerator();
+      Integer model_x = valX.getConst<Rational>().getNumerator();
+      Integer model_y = valY.getConst<Rational>().getNumerator();
+
+      if (TraceIsOn("piand-check"))
+      {
+        Trace("piand-check")
+            << "* " << i << ", value = " << valAndXY << std::endl;
+        Trace("piand-check") << "  actual (" << valX << ", " << valY
+                             << ") = " << valAndXYC << std::endl;
+      }
+      if (valAndXY == valAndXYC)
+      {
+        Trace("piand-check") << "...already correct" << std::endl;
+        continue;
+      }
+
+      Node lem_sum = sumBasedLemma(i, Kind::GEQ);
+      d_im.addPendingLemma(
```

Comment:
I don't know how or where, but i think this lemma should be added last. In iand it's the only one at this stage.